### PR TITLE
Stockfish train refactored

### PR DESCRIPTION
**Pull request by ryPattillo**
Starting the pull request, that way we can all look over it and make changes.

### CODE REVIEW

File: chess-AI/train.py [hunk 1]
```diff
@@ -1,32 +1,46 @@
...
+"""
+Main training program
+"""
+import time
 import os
+import chess
 import torch
-import numpy as np
+
 from torch.utils.data import DataLoader, TensorDataset
```
Comment (nashirj) [inline chess-AI/train.py:9]: `chess` and `torch` are not built-ins, so they should be grouped with the other torch include.
```py
import os
import time

import chess
import torch
from torch.utils.data import DataLoader, TensorDataset
```

File: chess-AI/train.py [hunk 2]
```diff
@@ -1,32 +1,46 @@
...
 from state_representation import get_cnn_input
-
+from stockfish_train import StockfishTrain
+from streamlit_dashboard import StreamlitDashboard
 
 class Train:
-    """This class is used to run the monte carlo simulations and drive the model training.
+    """The main training class, will train using stockfish and mcts
+    
+    1. Will create a dataset or load a dataset of (fen_strings,move_probs,values)
```
Comment (nashirj) [inline chess-AI/train.py:22]: You say create or load; when will a dataset be loaded? If it is an optional thing, maybe we need to add some parameter to the constructor to load the data.

File: chess-AI/train.py [hunk 3]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
     def assign_rewards(self, board, length):
         """Iterates through training examples and assigns rewards based on result of the game.
+        NOTE: Want to triple check this logic, reward assignment is very important
         """
         reward = 0
-        values = [0 for _ in range(length)]
+        
+        state_values = [0 for _ in range(length)]
 
+        # If board.outcome() is none , then there is a draw
```
Comment (nashirj) [inline chess-AI/train.py:117]:
```py
# if board.outcome() is None, the game is not over
# if board.outcome().winner is None, the game is a draw
```

File: chess-AI/train.py [hunk 4]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
 
-        # For demonstration print the board and outcome
-        # print(board)
-        print(values)
-        print(board.outcome())
-        return values
+    def create_dataset(self, games):
+        """Builds a dataset with the size of (games)
+
+        Parameters:
```
Comment (nashirj) [inline chess-AI/train.py:129]: Should be `Attributes` instead of `Parameters` for consistency with the other code, and `games` should be indented.

File: chess-AI/train.py [hunk 5]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
+        games: The total number of games to generate for the dataset
 
-    def training_episode(self, nnet, games, epochs, batch_size, num_saved_models, overwrite_save):
-        """Builds dataset from given number of training games and current network,
-        then trains the network on the MCTS output and game outcomes.
         """
 
+        # Using tensors but do not want to keep track of gradients
```
Comment (nashirj) [inline chess-AI/train.py:134]: Can you please explain here instead *why* we don't keep track of gradients

File: chess-AI/train.py [hunk 6]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
+        move_probs = torch.tensor([move_prob for game in game_data for move_prob in game[2]]).float()
+
+        # Create iterable dataset from game data
+        dataset = TensorDataset(input_state, state_values, move_probs)
+
+        # Return the dataset to be used
+        return dataset
+
+    def train_on_dataset(self, dataset, dashboard, nnet, epochs, batch_size, num_saved_models, overwrite_save):
+        """Function to train the with a dataset passed as a parameter
```
Comment (nashirj) [inline chess-AI/train.py:151]:
```py
"""Train using the specified `dataset`.
...
"""
```

File: chess-AI/train.py [hunk 7]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
+
+        Parameters:
```
Comment (nashirj) [inline chess-AI/train.py:153]: `Attributes`, and indent the attributes

File: chess-AI/train.py [hunk 8]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
+        # Stores the average losses which are used for graphing
+        average_pol_loss = []
+        average_val_loss = []
+
+        # Define loss functions (cross entropy and mean squared error)
+        ce_loss_fn = torch.nn.CrossEntropyLoss()
+        mse_loss_fn = torch.nn.MSELoss()
+
+        # Create optimizer for updating parameters during training
+        # NOTE: We are using SGD right now, maybe consider other optimizers such as Adam
```
Comment (nashirj) [inline chess-AI/train.py:170]: `# TODO: Consider using other optimizers, such as Adam`

File: chess-AI/train.py [hunk 9]
```diff
@@ -80,126 +100,234 @@ def training_game(self):
...
 
 def main():
-    mcts_simulations = 3
+    """Main function that will be run when starting training
+    """
+
+    # Detect device to train on
+    device = torch.device('cpu')
+    if torch.cuda.is_available():
+        device = torch.device('cuda:0')
```
Comment (nashirj) [inline chess-AI/train.py:255]:
```py
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
```

File: chess-AI/stockfish_train.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+import random
+import numpy as np
+from stockfish import Stockfish
+import chess
+
+from output_representation import PlayNetworkPolicyConverter
+
+
+class StockfishTrain:
+    """Class with all the necessary functionialities for
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:11]:
```py
"""Build a dataset of moves and stockfish evaluations.

This dataset is used to decrease training time for the Knightr0 AI. After initializing the
weights of the neural network using the Stockfish evaluations, we use self-play to
complete training.

...
"""
```

File: chess-AI/stockfish_train.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    """
+
+    def __init__(self, path):
+        self.stockfish = Stockfish(path)
+        self.policy_converter = PlayNetworkPolicyConverter()
+
+    def set_params(self, dashboard):
+        """ Sets the elo and depth for stockfish using the dashboard
+
+        Parameters:
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:26]: `Attributes` and indent below

Comment (nashirj) [reply chess-AI/stockfish_train.py:26]: This same comment applies to all the doc strings in this file.

File: chess-AI/stockfish_train.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        value: The value to take the sigmoid of
+        scale: How much to scale the value by
+        """
+
+        return 1 / (1 + np.exp(value/scale))
+
+    def choose_move(self, moves):
+        """ Chooses the next move to make
+        We want to make sure the top move isn't always what is
+        chosen
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:47]:
```py
"""Choose the next move to make with an epsilon-greedy strategy.

To encourage exploration, a random move is chosen with probability 1-epsilon.
Otherwise, the best action (according to the stockfish evaluation) is chosen.
"""

File: chess-AI/stockfish_train.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+        Parameters:
+        moves: The list of posssible moves to take
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:50]: `A list of legal moves, ordered by Stockfish evaluation from best to worst`

File: chess-AI/stockfish_train.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        """
+
+        epsilon = 0.3
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:53]: This should be passed as an argument to the function (it is a hyper parameter we may want to change later as one of the training options).

File: chess-AI/stockfish_train.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        
+        # Return a random number in the range (0,1)
+        rand = np.random.rand()
+
+        # Choose the best move 30% of the time
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:58]: `# With probability 1-epsilon, choose the best move`

File: chess-AI/stockfish_train.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        if rand < epsilon:
+            move = moves[0]
+        # Choose a random move 70% of the time , allows for exploration
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:61]: `# With probability epsilon, choose a random move (to encourage exploration).`

File: chess-AI/stockfish_train.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        else:
+            move = moves[random.randint(0, len(moves) - 1)]
+
+        return move
+
+    def get_value(self, board):
+        """ Returns a value for a given state on the board
+        Utilizes stockfish Centipawn calculation through stockfish.get_evaluation()
+        Will then use sig() to transform the value between (0,1)
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:70]:
```py
"""Return a Stockfish evaluation of the given position.

The optional `sig` parameter specifies whether to softmax
the return value (by default, the value is softmaxed).
"""
```

File: chess-AI/stockfish_train.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+        # get_evaluation() always from white perspective, need to account for that
+        # Positive is advantage white, negative is advantage black
+
+        player = 1 if board.turn == chess.WHITE else -1
+        
+        # Player will transform the value depending on if black or white
+        state_value = self.stockfish.get_evaluation()["value"] * player
+
+        # Will use the sig() and transform between (0,1)
```
Comment (nashirj) [inline chess-AI/stockfish_train.py:81]: `# Transform state value to be in the range [-1, 1]`

**Review by nashirj — COMMENTED** [review]
## chess-AI/datasets/stockfish_data.pt
I don't think we should store all this data on github, it's going to change after everytime you run training, and I can't see why we would need to store old versions of the data. If I'm misunderstanding and there's a good reason to store old data lmk. Otherwise, I think we should do the following:
```bash
# remove the file
git rm --cached <file>
# commit with amend flag
git commit --amend
```
Since we added a `.gitignore` in #136, we won't have to worry about future `*.pt` files after making this fix.

## documentation.md
I think the documentation should explain data flow and how pieces of the program instead of explaining what each function does. The explanation should be comments in the code (like more thorough docstrings maybe), but the documentation should explain how everything fits together and how the whole program flows. These explanations are great but should be included in the code, whereas the docs should help with a high level understanding of how everything works
